modal/io_streams.py

```diff
@@ -251,21 +252,27 @@ async def drain(self):
         self._buffer.clear()
         index = self.get_next_index()
 
-        if self._object_type == "sandbox":
-            await retry_transient_errors(
-                self._client.stub.SandboxStdinWrite,
-                api_pb2.SandboxStdinWriteRequest(
-                    sandbox_id=self._object_id, index=index, eof=self._is_closed, input=data
-                ),
-            )
-        else:
-            await retry_transient_errors(
-                self._client.stub.ContainerExecPutInput,
-                api_pb2.ContainerExecPutInputRequest(
-                    exec_id=self._object_id,
-                    input=api_pb2.RuntimeInputMessage(message=data, message_index=index, eof=self._is_closed),
-                ),
-            )
+        try:
+            if self._object_type == "sandbox":
+                await retry_transient_errors(
+                    self._client.stub.SandboxStdinWrite,
+                    api_pb2.SandboxStdinWriteRequest(
+                        sandbox_id=self._object_id, index=index, eof=self._is_closed, input=data
+                    ),
+                )
+            else:
+                await retry_transient_errors(
+                    self._client.stub.ContainerExecPutInput,
+                    api_pb2.ContainerExecPutInputRequest(
+                        exec_id=self._object_id,
+                        input=api_pb2.RuntimeInputMessage(message=data, message_index=index, eof=self._is_closed),
+                    ),
+                )
+        except GRPCError as exc:
+            if exc.status == Status.FAILED_PRECONDITION:
+                raise EOFError("Stdin is closed. Cannot write to it.")
```

Comment:
Can't you grab the message in the gRPC exception? It'd be more accurate and forward compatible. 
Also not sure `EOFError` is what a user should expect here, could we instead throw an exception from within the `modal.exception` namespace that is like 'invalid error'?

Also re-reading the docstring I don't like that it suggests that `drain()` may always send EOF. It'd be nice to reword that to indicate EOF is only sent once the writer is _closed_.

Reply:
Fixed the comments and switched to ValueError to match python native socket behavior.